PCHR-2429: Fix unassigned records in admin leave requests

igorpavlov opened the pull request:
## Overview

For admin on Admin Dashboard (or Manager Leave) pages, unassigned contacts do not appear in "All" filter list. This PR fixes this issue.

## Before

![1](https://user-images.githubusercontent.com/3973243/28582224-93f8b340-715c-11e7-9412-5304fac6a11e.gif)

## After

![2](https://user-images.githubusercontent.com/3973243/28582231-978d3530-715c-11e7-8886-a2593b7fe20a.gif)

## Technical Details

Filtering function was sending `unassigned` as `false`, but API expected `unassigned` to be `undefined` to return unassigned contacts as well.

## Comments

1. This PR also partially refactors tests for Manage Leave Requests component.
2. The following commit https://github.com/civicrm/civihr/pull/2023/commits/9fdc3243118a17acde9793fd99b3251710de397d unwraps the code from global CRM variable. It was made as a single commit for better picture of visual changes. The real refactoring is in the commit https://github.com/civicrm/civihr/pull/2023/commits/c173340e2ce3fa9e4d94516a11269752a2655718

---

- [x] Tests Pass

reneolivo commented on uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/manage-leave-requests.spec.js:
```diff
@@ -1,929 +1,953 @@
-              label: 'All'
-            });
+    describe('data loading', function () {
+      // TODO need to figure out how to test variables which are changed when controller gets initialized
```
Hint: use the `$onInit` method! Not sure how to apply it for this case though.

But the gist is to use the constructor to setup the controller before initializing it, you test these settings on your spec file, then run `$onInit` manually and test the initialization effects.

Note: this is just a comment.

reneolivo commented on uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/manage-leave-requests.spec.js:
```diff
@@ -1,929 +1,953 @@
-          });
+        it('leave requests have loaded', function () {
+          expect(controller.leaveRequests.table.list.length).toEqual(leaveRequestData.all().values.length);
+          expect(controller.leaveRequests.filter.list.length).toEqual(leaveRequestData.all().values.length);
```
It's better to have each expect in a different `it` statement in case one fails it's makes finding the error easier.

This one could be something like `it('updates filters for each type of leave request loaded', function() {` or at least that's what I think it's doing.

This also helps explain what's supposed to happen (if anyone reads the specs for that).

igorpavlov replied:
Good point. I suggest to use `loaded leave requests data for filters` because that's what it exactly does.

reneolivo commented on uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/manage-leave-requests.spec.js:
```diff
@@ -1,929 +1,953 @@
-                    location: null
-                  }));
-                });
+              it('Contact API gets called with null', function () {
```
it calls Contact API with a null value

reneolivo commented on uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/manage-leave-requests.spec.js:
```diff
@@ -1,929 +1,953 @@
-                    level_type: mockLevelType
-                  }));
-                });
+              it('Contact API gets called with location value', function () {
```
it calls Contact API with the location value

reneolivo commented on uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/manage-leave-requests.spec.js:
```diff
@@ -1,929 +1,953 @@
-                    level_type: null
-                  }));
-                });
+              it('ContactAPI gets called with null', function () {
```
same.

reneolivo commented on uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/manage-leave-requests.spec.js:
```diff
@@ -1,929 +1,953 @@
-                    region: mockRegion
-                  }));
-                });
+              it('ContactAPI gets called with level types value', function () {
```
same

reneolivo commented on uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/manage-leave-requests.spec.js:
```diff
@@ -1,929 +1,953 @@
-                    region: null
-                  }));
-                });
+              it('Contact API gets called with null', function () {
```
same.

ummmmmm, are we testing the same thing over and over?

Please update the wording wherever it appears (there are still a few more to go).

igorpavlov commented:
@reneolivo I have refactored tests even more:
1) I wrapped repeating parts in functions. We can wrap more, but I don't want to wrap `describe` and `it` parts to be honest because in the case of test failing it would be difficult to debug.
2) I renamed `it`s as you suggested
3) as per the `before loading starts` section I decided to leave it, it doesn't seem possible to make such tests without changes in the controller and since we follow JS style guidelines we should not change it until we discuss this with the team.

reneolivo commented:
All changed are good. Feel free to merge.
